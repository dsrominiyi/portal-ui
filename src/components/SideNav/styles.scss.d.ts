export const sideNav: string;
export const mini: string;
export const headerContainer: string;
export const divider: string;
export const toggleButton: string;
