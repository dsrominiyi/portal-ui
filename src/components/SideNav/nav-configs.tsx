import React from 'react';
import { NavItem } from '.';

import CoversureLogoText from '../../assets/svg/logo-coversure-text.svg';
import CoversureLogoArrows from '../../assets/svg/logo-coversure-arrows.svg';

import routes from '../../routes';

const links = {
  dashboard: {
    icon: <i className="fas fa-ellipsis-h" />,
    label: 'Dashboard',
    href: routes.dashboard
  },
  customer: {
    icon: <i className="fas fa-users" />,
    label: 'Customer',
    subItems: [
      { label: 'New Individual', href: routes.newIndividual },
      { label: 'New Company', href: routes.newCompany },
      { label: 'Find Customer' }
    ]
  },
  customerOverview: { icon: <i className="fas fa-user" />, label: 'Customer Overview' },
  customerDetails: { icon: <i className="fas fa-user-edit" />, label: 'Customer Details' },
  diary: { icon: <i className="fas fa-calendar-alt" />, label: 'Diary' },
  renewals: { icon: <i className="fas fa-file-alt" />, label: 'Renewals' },
  officeManagement: { icon: <i className="fas fa-sliders-h" />, label: 'Office Management' },
  payments: { icon: <i className="fas fa-credit-card" />, label: 'Payments' },
  reports: { icon: <i className="fas fa-file-pdf" />, label: 'Reports' },
  messages: { icon: <i className="fas fa-comments" />, label: 'Messages' },
  complaints: { icon: <i className="fas fa-frown" />, label: 'Complaints' },
  settings: { icon: <i className="fas fa-cog" />, label: 'Settings' },
  policies: { icon: <i className="fas fa-folder" />, label: 'Policies' },
  quotes: { icon: <i className="fas fa-folder-open" />, label: 'Quotes', href: routes.quotes },
  crossSell: { icon: <i className="fas fa-link" />, label: 'Cross Sell' },
  history: { icon: <i className="fas fa-history" />, label: 'History' },
  letters: { icon: <i className="fas fa-envelope-open-text" />, label: 'Letters' }
};

const generalLinks: NavItem[] = [
  links.diary,
  links.renewals,
  links.officeManagement,
  links.payments,
  links.reports,
  links.messages,
  links.complaints,
  links.settings
];

const header = <CoversureLogoText />;
const miniHeader = <CoversureLogoArrows />;

// configs -------------------------------

export const dashboard = {
  header,
  miniHeader,
  items: [{ ...links.dashboard, isActive: true }, links.customer, '-', ...generalLinks] as NavItem[]
};

export const customer = {
  header,
  miniHeader,
  items: [links.dashboard, { ...links.customer, isActive: true }, '-', ...generalLinks] as NavItem[]
};

export const customerOverview = {
  header,
  miniHeader,
  items: [
    links.dashboard,
    links.customer,
    { ...links.customerOverview, isActive: true },
    links.customerDetails,
    links.policies,
    links.quotes,
    links.crossSell,
    links.history,
    links.letters,
    '-',
    ...generalLinks
  ] as NavItem[]
};
