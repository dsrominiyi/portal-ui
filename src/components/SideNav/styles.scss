@import '../../style/vars';

.side-nav {
  display: flex;
  flex-direction: column;
  flex: 0 0 224px;
  width: 224px;
  background-color: $brand-dark-blue;
  transition: flex 0.3s, width 0.3s;

  &.mini {
    flex: 0 0 104px;
    width: 104px;
  }
}

.header-container {
  padding: 1.5rem;
}

.divider {
  margin: 1rem;
  border-bottom: 1px lighten($brand-dark-blue, 10%) solid;
}

.toggle-button {
  display: flex;
  justify-content: center;
  padding: 1rem;
}
