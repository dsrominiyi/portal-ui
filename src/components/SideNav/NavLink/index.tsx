import React, { FunctionComponent, ReactNode, MouseEvent } from 'react';
import classnames from 'classnames/bind';

import Link from '../../Link';

import styles from './styles.scss';

interface Props {
  label: string;
  icon: ReactNode;
  href?: string;
  isActive?: boolean;
  isMiniLink?: boolean;
  onClick?(event: MouseEvent): void;
}

const cx = classnames.bind(styles);

export const NavLink: FunctionComponent<Props> = ({
  label,
  icon,
  href,
  isActive,
  isMiniLink,
  onClick
}: Props) => (
  <Link
    classNames={[
      cx({
        navLink: true,
        active: isActive,
        mini: isMiniLink
      })
    ]}
    label={label}
    icon={icon}
    href={href}
    onClick={onClick}
  />
);

NavLink.displayName = 'NavLink';
export default NavLink;
