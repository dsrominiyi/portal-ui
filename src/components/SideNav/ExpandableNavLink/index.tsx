import React, {
  FunctionComponent,
  ReactNode,
  useState,
  useRef,
  useEffect,
  MouseEvent
} from 'react';
import classnames from 'classnames/bind';
import Link from '../../Link';
import NavLink from '../NavLink';
import styles from './styles.scss';

const cx = classnames.bind(styles);

export interface SubItem {
  label: string;
  href?: string;
  onClick?(event: MouseEvent): void;
}

export interface Props {
  label: string;
  icon: ReactNode;
  isActive?: boolean;
  isMiniLink?: boolean;
  subItems: SubItem[];
}

export const ExpandableNavLink: FunctionComponent<Props> = ({
  label,
  icon,
  isActive,
  isMiniLink,
  subItems
}: Props) => {
  const [isExpanded, toggleExpand] = useState(false);
  const dropdownRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    const dropdownDiv = dropdownRef.current as HTMLInputElement;
    if (!isMiniLink) {
      dropdownDiv.style.height = isExpanded ? `${dropdownDiv.scrollHeight}px` : '0';
    } else {
      dropdownDiv.style.height = 'auto';
    }
  }, [isExpanded, isMiniLink]);

  useEffect(() => {
    if (isExpanded) {
      toggleExpand(!isExpanded);
    }
  }, [isMiniLink]);

  const displayLinks = subItems.map((subItem, index) => {
    return (
      <Link
        key={index}
        label={subItem.label}
        href={subItem.href}
        onClick={subItem.href ? undefined : subItem.onClick}
        classNames={[styles.link]}
      />
    );
  });

  return (
    <div className={styles.expandableNavLink}>
      <div
        className={cx({ clickableLink: true, mini: isMiniLink })}
        onClick={() => toggleExpand(!isExpanded)}
      >
        <NavLink label={label} icon={icon} isActive={isActive} isMiniLink={isMiniLink} />
        {!isMiniLink && (
          <i
            className={cx({ chevron: true, 'fas fa-chevron-right': true, expanded: isExpanded })}
          />
        )}
      </div>
      <div
        ref={dropdownRef}
        className={cx({
          dropDown: true,
          mini: isMiniLink,
          expanded: !isMiniLink && isExpanded,
          hidden: isMiniLink && !isExpanded
        })}
      >
        {displayLinks}
      </div>
    </div>
  );
};

ExpandableNavLink.displayName = 'ExpandableNavLink';
export default ExpandableNavLink;
