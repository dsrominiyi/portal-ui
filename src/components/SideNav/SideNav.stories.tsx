import React from 'react';
import { storiesOf } from '@storybook/react';

import SideNav, { Props as SideNavProps } from '.';

import * as configs from './nav-configs';

const stories = storiesOf('SideNav', module);

const render = (props?: SideNavProps) => () => <SideNav {...props} />;

export const storyRenders = {
  dashboardConfig: render(configs.dashboard),
  customerOverviewConfig: render(configs.customerOverview),
  noItems: render(),
  invalidItems: render({ items: [{}, '*', 'foo'] } as any)
};

stories.add('Dashboard config', storyRenders.dashboardConfig);
stories.add('Customer config', storyRenders.customerOverviewConfig);
stories.add('No items', storyRenders.noItems);
stories.add('Invalid items', storyRenders.invalidItems);
