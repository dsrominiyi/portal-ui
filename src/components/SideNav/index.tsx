import React, { FunctionComponent, MouseEvent, ReactNode, useState } from 'react';
import classnames from 'classnames/bind';

import ChevronLeft from 'uikit-icons/lib/ChevronLeft';
import ChevronRight from 'uikit-icons/lib/ChevronRight';

import NavLink from './NavLink';
import ExpandableNavLink, { SubItem } from './ExpandableNavLink';
import CircularButton from '../Button/CircularButton';

import styles from './styles.scss';

interface SingleLink {
  icon: ReactNode;
  label: string;
  href?: string;
  isActive?: boolean;
  onClick?(event: MouseEvent): void;
}

interface ExpandableLink {
  icon: ReactNode;
  label: string;
  isActive?: boolean;
  subItems: SubItem[];
}

type Divider = '-';

export type NavItem = SingleLink | ExpandableLink | Divider;

export interface Props {
  items?: NavItem[];
  header?: ReactNode;
  miniHeader?: ReactNode;
}

const cx = classnames.bind(styles);

const isSingleLink = (item: NavItem): item is SingleLink => {
  return (
    typeof item !== 'string' &&
    typeof (item as SingleLink).label === 'string' &&
    (item as ExpandableLink).subItems === undefined
  );
};

const isExpandableLink = (item: NavItem): item is ExpandableLink => {
  return typeof item !== 'string' && (item as ExpandableLink).subItems !== undefined;
};

const isDivider = (item: NavItem): item is Divider => item === '-';

export const SideNav: FunctionComponent<Props> = ({ items = [], header, miniHeader }: Props) => {
  const [isMiniNav, toggleMiniNav] = useState(false);

  const renderedItems = items.map((item, index) => {
    if (isSingleLink(item)) {
      return <NavLink key={item.label} {...item} isMiniLink={isMiniNav} />;
    }
    if (isExpandableLink(item)) {
      return <ExpandableNavLink key={item.label} {...item} isMiniLink={isMiniNav} />;
    }
    if (isDivider(item)) {
      return <div key={index} className={styles.divider} />;
    }
    return undefined;
  });

  const showHeader = (header && !isMiniNav) || (miniHeader && isMiniNav);

  return (
    <div className={cx({ sideNav: true, mini: isMiniNav })}>
      {showHeader && (
        <div className={styles.headerContainer}>{isMiniNav ? miniHeader : header}</div>
      )}
      {renderedItems}
      <div className={styles.toggleButton}>
        <CircularButton
          label={isMiniNav ? <ChevronRight /> : <ChevronLeft />}
          onClick={() => toggleMiniNav(!isMiniNav)}
        />
      </div>
    </div>
  );
};

SideNav.displayName = 'SideNav';
export default SideNav;
