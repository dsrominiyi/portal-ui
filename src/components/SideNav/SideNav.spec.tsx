import React from 'react';
import { shallow } from 'enzyme';
import { storyRenders } from './SideNav.stories';

import SideNav from '.';
import { dashboard as dashboardConfig } from './nav-configs';

import testStorySnapshots from '../../helpers/test/testStorySnapshots';

describe('SideNav', () => {
  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('clicking the CircularButton toggles the isMiniNav state value', () => {
    const props = {
      header: <div className="header" />,
      miniHeader: <div className="miniHeader" />,
      items: dashboardConfig.items
    };
    const component = shallow(<SideNav {...props} />);

    // isMiniNav will be true
    component.find('CircularButton').simulate('click');

    expect(component.find('.header').length).toBe(0);
    expect(component.find('.miniHeader').length).toBe(1);
    component.find('NavLink').forEach(navLink => {
      expect((navLink.props() as any).isMiniLink).toBe(true);
    });
    component.find('ExpandableNavLink').forEach(navLink => {
      expect((navLink.props() as any).isMiniLink).toBe(true);
    });

    // isMiniNav will be false
    component.find('CircularButton').simulate('click');

    expect(component.find('.header').length).toBe(1);
    expect(component.find('.miniHeader').length).toBe(0);
    component.find('NavLink').forEach(navLink => {
      expect((navLink.props() as any).isMiniLink).toBe(false);
    });
    component.find('ExpandableNavLink').forEach(navLink => {
      expect((navLink.props() as any).isMiniLink).toBe(false);
    });
  });
});
