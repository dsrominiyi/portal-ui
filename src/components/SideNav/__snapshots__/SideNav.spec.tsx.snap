// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav component story renders match snapshots: customerOverviewConfig 1`] = `
<div
  className="sideNav"
>
  <div
    className="headerContainer"
  >
    <svg />
  </div>
  <NavLink
    href="/dashboard"
    icon={
      <i
        className="fas fa-ellipsis-h"
      />
    }
    isMiniLink={false}
    key="Dashboard"
    label="Dashboard"
  />
  <ExpandableNavLink
    icon={
      <i
        className="fas fa-users"
      />
    }
    isMiniLink={false}
    key="Customer"
    label="Customer"
    subItems={
      Array [
        Object {
          "href": "/new-individual",
          "label": "New Individual",
        },
        Object {
          "href": "/new-company",
          "label": "New Company",
        },
        Object {
          "label": "Find Customer",
        },
      ]
    }
  />
  <NavLink
    icon={
      <i
        className="fas fa-user"
      />
    }
    isActive={true}
    isMiniLink={false}
    key="Customer Overview"
    label="Customer Overview"
  />
  <NavLink
    icon={
      <i
        className="fas fa-user-edit"
      />
    }
    isMiniLink={false}
    key="Customer Details"
    label="Customer Details"
  />
  <NavLink
    icon={
      <i
        className="fas fa-folder"
      />
    }
    isMiniLink={false}
    key="Policies"
    label="Policies"
  />
  <NavLink
    href="/customer/quotes"
    icon={
      <i
        className="fas fa-folder-open"
      />
    }
    isMiniLink={false}
    key="Quotes"
    label="Quotes"
  />
  <NavLink
    icon={
      <i
        className="fas fa-link"
      />
    }
    isMiniLink={false}
    key="Cross Sell"
    label="Cross Sell"
  />
  <NavLink
    icon={
      <i
        className="fas fa-history"
      />
    }
    isMiniLink={false}
    key="History"
    label="History"
  />
  <NavLink
    icon={
      <i
        className="fas fa-envelope-open-text"
      />
    }
    isMiniLink={false}
    key="Letters"
    label="Letters"
  />
  <div
    className="divider"
    key="9"
  />
  <NavLink
    icon={
      <i
        className="fas fa-calendar-alt"
      />
    }
    isMiniLink={false}
    key="Diary"
    label="Diary"
  />
  <NavLink
    icon={
      <i
        className="fas fa-file-alt"
      />
    }
    isMiniLink={false}
    key="Renewals"
    label="Renewals"
  />
  <NavLink
    icon={
      <i
        className="fas fa-sliders-h"
      />
    }
    isMiniLink={false}
    key="Office Management"
    label="Office Management"
  />
  <NavLink
    icon={
      <i
        className="fas fa-credit-card"
      />
    }
    isMiniLink={false}
    key="Payments"
    label="Payments"
  />
  <NavLink
    icon={
      <i
        className="fas fa-file-pdf"
      />
    }
    isMiniLink={false}
    key="Reports"
    label="Reports"
  />
  <NavLink
    icon={
      <i
        className="fas fa-comments"
      />
    }
    isMiniLink={false}
    key="Messages"
    label="Messages"
  />
  <NavLink
    icon={
      <i
        className="fas fa-frown"
      />
    }
    isMiniLink={false}
    key="Complaints"
    label="Complaints"
  />
  <NavLink
    icon={
      <i
        className="fas fa-cog"
      />
    }
    isMiniLink={false}
    key="Settings"
    label="Settings"
  />
  <div
    className="toggleButton"
  >
    <CircularButton
      label={<SvgChevronLeft />}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`SideNav component story renders match snapshots: dashboardConfig 1`] = `
<div
  className="sideNav"
>
  <div
    className="headerContainer"
  >
    <svg />
  </div>
  <NavLink
    href="/dashboard"
    icon={
      <i
        className="fas fa-ellipsis-h"
      />
    }
    isActive={true}
    isMiniLink={false}
    key="Dashboard"
    label="Dashboard"
  />
  <ExpandableNavLink
    icon={
      <i
        className="fas fa-users"
      />
    }
    isMiniLink={false}
    key="Customer"
    label="Customer"
    subItems={
      Array [
        Object {
          "href": "/new-individual",
          "label": "New Individual",
        },
        Object {
          "href": "/new-company",
          "label": "New Company",
        },
        Object {
          "label": "Find Customer",
        },
      ]
    }
  />
  <div
    className="divider"
    key="2"
  />
  <NavLink
    icon={
      <i
        className="fas fa-calendar-alt"
      />
    }
    isMiniLink={false}
    key="Diary"
    label="Diary"
  />
  <NavLink
    icon={
      <i
        className="fas fa-file-alt"
      />
    }
    isMiniLink={false}
    key="Renewals"
    label="Renewals"
  />
  <NavLink
    icon={
      <i
        className="fas fa-sliders-h"
      />
    }
    isMiniLink={false}
    key="Office Management"
    label="Office Management"
  />
  <NavLink
    icon={
      <i
        className="fas fa-credit-card"
      />
    }
    isMiniLink={false}
    key="Payments"
    label="Payments"
  />
  <NavLink
    icon={
      <i
        className="fas fa-file-pdf"
      />
    }
    isMiniLink={false}
    key="Reports"
    label="Reports"
  />
  <NavLink
    icon={
      <i
        className="fas fa-comments"
      />
    }
    isMiniLink={false}
    key="Messages"
    label="Messages"
  />
  <NavLink
    icon={
      <i
        className="fas fa-frown"
      />
    }
    isMiniLink={false}
    key="Complaints"
    label="Complaints"
  />
  <NavLink
    icon={
      <i
        className="fas fa-cog"
      />
    }
    isMiniLink={false}
    key="Settings"
    label="Settings"
  />
  <div
    className="toggleButton"
  >
    <CircularButton
      label={<SvgChevronLeft />}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`SideNav component story renders match snapshots: invalidItems 1`] = `
<div
  className="sideNav"
>
  <div
    className="toggleButton"
  >
    <CircularButton
      label={<SvgChevronLeft />}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`SideNav component story renders match snapshots: noItems 1`] = `
<div
  className="sideNav"
>
  <div
    className="toggleButton"
  >
    <CircularButton
      label={<SvgChevronLeft />}
      onClick={[Function]}
    />
  </div>
</div>
`;
